porch/controllers/rootsyncsets/pkg/controllers/rootsyncset/controller.go

```diff
@@ -108,59 +116,215 @@ func (r *RootSyncSetReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		// Stop reconciliation as the item is being deleted
 		return ctrl.Result{}, nil
 	}
-	var patchErrs []error
+
+	results := make(reconcileResult)
 	for _, clusterRef := range rootsyncset.Spec.ClusterRefs {
 		clusterRefName := clusterRef.Kind + ":" + clusterRef.Name
+		result := clusterRefResult{}
 		client, err := r.GetClient(ctx, clusterRef, rootsyncset.Namespace)
 		if err != nil {
-			patchErrs = append(patchErrs, err)
-			continue
-		}
-		rootSyncRes, newRootSync, err := BuildObjectsToApply(&rootsyncset)
-		if err != nil {
-			patchErrs = append(patchErrs, err)
+			result.clientError = err
 			continue
 		}
-		data, err := json.Marshal(newRootSync)
-		if err != nil {
-			patchErrs = append(patchErrs, fmt.Errorf("failed to encode root sync to JSON: %w", err))
-			continue
+
+		if err := r.patchRootSync(ctx, client, req.Name, &rootsyncset); err != nil {
+			result.patchError = err
 		}
-		rs, err := client.Resource(rootSyncRes).Namespace(rootSyncNamespace).Patch(ctx, req.Name, types.ApplyPatchType, data, metav1.PatchOptions{FieldManager: req.Name})
+
+		s, err := r.checkSyncStatus(ctx, client, req.Name)
 		if err != nil {
-			patchErrs = append(patchErrs, fmt.Errorf("failed to patch RootSync %s in cluster %s: %w", rootSyncNamespace+"/"+req.Name, clusterRefName, err))
+			result.statusError = err
+			result.status = "Unknown"
 		} else {
-			klog.Infof("Create/Update resource %s as %v", req.Name, rs)
+			result.status = s
 		}
+
+		results[clusterRefName] = result
 	}
-	if len(patchErrs) != 0 {
-		for _, patchErr := range patchErrs {
-			klog.Errorf("%v", patchErr)
+	for clusterRef, res := range results {
+		if res.clientError != nil {
+			klog.Errorf("failed to create client for %s: %v", clusterRef, res.clientError)
+		}
+		if res.patchError != nil {
+			klog.Errorf("failed to patch %s: %v", clusterRef, res.patchError)
 		}
-		return ctrl.Result{}, patchErrs[0]
+		if res.statusError != nil {
+			klog.Errorf("failed to check status for %s: %v", clusterRef, res.statusError)
+		}
+	}
+
+	if errs := results.Errors(); len(errs) > 0 {
+		return ctrl.Result{}, errs[0]
 	}
 	return ctrl.Result{}, nil
 }
 
-// BuildObjectsToApply config root sync
-func BuildObjectsToApply(rootsyncset *v1alpha1.RootSyncSet) (schema.GroupVersionResource, *unstructured.Unstructured, error) {
-	gv, err := schema.ParseGroupVersion(rootSyncApiVersion)
+type reconcileResult map[string]clusterRefResult
```

Comment:
We could also have this implement `error`

Reply:
Yeah, added that.